File: management-domain/Finance-department/Backend/Java/accounts-payable-service/src/main/java/com/gogidix/finance/accountspayable/domain/event/InvoiceApprovedEvent.java
package com.gogidix.finance.accountspayable.domain.event;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.UUID;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class InvoiceApprovedEvent {

    private String eventId;
    private String invoiceId;
    private String tenantId;
    private String vendorId;
    private String invoiceNumber;
    private BigDecimal amount;
    private String currency;
    private String approver;
    private String approvalLevel;
    private Instant timestamp;
    private String eventType;

    public static InvoiceApprovedEvent create(String invoiceId, String tenantId, String vendorId,
                                               String invoiceNumber, BigDecimal amount, String currency,
                                               String approver, String approvalLevel) {
        return InvoiceApprovedEvent.builder()
            .eventId(UUID.randomUUID().toString())
            .invoiceId(invoiceId)
            .tenantId(tenantId)
            .vendorId(vendorId)
            .invoiceNumber(invoiceNumber)
            .amount(amount)
            .currency(currency)
            .approver(approver)
            .approvalLevel(approvalLevel)
            .timestamp(Instant.now())
            .eventType("INVOICE_APPROVED")
            .build();
    }
}

File: management-domain/Finance-department/Backend/Nodes/finance-automation-service/invoice-processing-service/src/domain/events/invoice-approved.event.ts
import { InvoiceStatus } from '../models/invoice-validation.entity';
import { Invoice, PaymentTerms } from '../models/invoice.entity';

/**
 * Domain event emitted when an invoice is approved
 */
export interface InvoiceApprovedEvent {
  eventType: 'InvoiceApproved';
  eventId: string;
  occurredAt: Date;
  tenantId: string;
  organizationId: string;
  correlationId?: string;
  payload: {
    invoiceId: string;
    invoiceNumber: string;
    vendorId: string;
    vendorName: string;
    totalAmount: number;
    currency: string;
    status: InvoiceStatus;
    approvedBy: string;
    approvedAt: Date;
    category?: string;
    glAccountCode?: string;
    costCenter?: string;
    paymentTerms: PaymentTerms;
    dueDate: Date;
    outstandingAmount: number;
  };
}

export class InvoiceApprovedEventFactory {
  static create(
    invoice: Invoice,
    correlationId?: string,
  ): InvoiceApprovedEvent {
    return {
      eventType: 'InvoiceApproved',
      eventId: this.generateEventId(),
      occurredAt: new Date(),
      tenantId: invoice.getTenantId(),
      organizationId: invoice.getOrganizationId(),
      correlationId,
      payload: {
        invoiceId: invoice.id?.toString() || '',
        invoiceNumber: invoice.getInvoiceNumber(),
        vendorId: invoice.getVendorId(),
        vendorName: invoice.getVendorName(),
        totalAmount: invoice.getTotalAmount(),
        currency: invoice.getCurrency(),
        status: invoice.getStatus(),
        approvedBy: invoice.getApprovedBy() || 'system',
        approvedAt: invoice.getApprovedAt() || new Date(),
        category: invoice.getCategory(),
        glAccountCode: invoice.getGlAccountCode(),
        costCenter: invoice.getCostCenter(),
        paymentTerms: invoice.getPaymentTerms(),
        dueDate: invoice.getDueDate(),
        outstandingAmount: invoice.getOutstandingAmount(),
      },
    };
  }

  private static generateEventId(): string {
    return `evt-${Date.now()}-${Math.random().toString(36).substring(2, 9)}`;
  }
}
